src/models/protocol/types.rs

```diff
@@ -0,0 +1,352 @@
+use serde::de::Visitor;
+use serde::{Deserialize, Deserializer};
+use serde_json::Value;
+use std::borrow::{Cow};
+use std::collections::HashMap;
+
+#[derive(Deserialize,Debug)]
+pub struct BitField {
+    pub name: String,
+    pub size: i64,
+    pub signed: bool,
+}
+
+/// These data types should be available in every version.
+/// However, they won't break anything if not present
+/// This can also be known as the Native Types
+#[derive(Debug)]
+pub enum NativeType {
+    /// Please read the following link for information on parsing https://wiki.vg/Protocol#VarInt_and_VarLong
+    VarInt,
+    PString {
+        count_type: Box<NativeType>,
+    },
+    Buffer {
+        count_type: Box<NativeType>,
+    },
+    Bool,
+    U8,
+    U16,
+    U32,
+    U64,
+    I8,
+    I16,
+    I32,
+    I64,
+    F32,
+    F64,
+    Uuid,
+    // Optional<MinecraftPacketDataType>
+    Option(Box<PacketDataType>),
+    EntityMetadataLoop {
+        end_val: i64,
+        metadata_type: Box<PacketDataType>,
+    },
+    TopBitSetTerminatedArray(Box<PacketDataType>),
+    BitField(Vec<BitField>),
+    // A set of Name and The Type
+    Container(Vec<(String, Box<PacketDataType>)>),
+    Switch {
+        compare_to: String,
+        fields: HashMap<String, String>,
+        default: Option<String>,
+    },
+    Void,
+    Array {
+        count_type: Box<NativeType>,
+        array_type: Box<PacketDataType>,
+    },
+    RestBuffer,
+    NBT,
+    OptionalNBT,
+}
+
+impl NativeType {
+    pub fn contains_type(name: &str) -> bool {
+        matches!(name, "varint" | "pstring" | "buffer" | "bool" | "u8" | "u16" | "u32" | "u64" | "i8" | "i16" | "i32" | "i64" | "f32" | "f64" | "uuid" | "option" | "entityMetadataLoop" | "topbitsetterminatedarray" | "bitfield" | "container" | "switch" | "void" | "array" | "restbuffer" | "nbt" | "optionalnbt")
+    }
+    pub fn new(name: &str, layout: Cow<'_, Value>) -> Option<Self> {
```

Comment:
I think this can be split up into functions like `new_varint`, `new_pstring` that get called in the match arms